programs/ssl/ssl_test_common_source.c

```diff
@@ -178,29 +179,34 @@ int ssl_check_record( mbedtls_ssl_context const *ssl,
             case MBEDTLS_ERR_SSL_INVALID_RECORD:
                 if( opt.debug_level > 1 )
                     mbedtls_printf( "mbedtls_ssl_check_record() detected invalid record.\n" );
+                ret = 0;
                 break;
 
             case MBEDTLS_ERR_SSL_INVALID_MAC:
                 if( opt.debug_level > 1 )
                     mbedtls_printf( "mbedtls_ssl_check_record() detected unauthentic record.\n" );
+                ret = 0;
                 break;
 
             case MBEDTLS_ERR_SSL_UNEXPECTED_RECORD:
                 if( opt.debug_level > 1 )
                     mbedtls_printf( "mbedtls_ssl_check_record() detected unexpected record.\n" );
+                ret = 0;
                 break;
 
             default:
                 mbedtls_printf( "mbedtls_ssl_check_record() failed fatally with -%#04x.\n", (unsigned int) -ret );
-                return( -1 );
+                ret = -1;
+                goto cleanup;
         }
 
         /* Regardless of the outcome, forward the record to the stack. */
     }
 
+cleanup:
     mbedtls_free( tmp_buf );
 
-    return( 0 );
+    return( ret );
```

Comment:
I wonder, shouldn't have this been caught by some test?
`MBEDTLS_ERR_SSL_FEATURE_UNAVAILABLE` value is  `-0x7080`
so a test case checking for return value of unsupported feature should expect a 0, receive -0x7800 and fail. Right?

Reply:
Good question!

Looking at the source, it appears that `MBEDTLS_ERR_SSL_FEATURE_UNAVAILABLE` is only returned when used with TLS (as opposed to DTLS), and looking at the places the function is called in `ssl_test_common_source.c` it appears we're only calling it when using DTLS. So I think it's OK that this wasn't caught. Arguably the test code is a bit redundant here, which I think is OK.